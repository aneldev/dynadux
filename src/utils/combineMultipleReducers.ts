import { IDynaduxReducerDic } from "../Dynadux/Dynadux";


export const combineMultipleReducers = <TState>(...reducerDics: Array<IDynaduxReducerDic<TState>>): IDynaduxReducerDic<TState> => {
  const output: IDynaduxReducerDic<TState> = {};

  reducerDics
    .forEach(reducerDic => {
      Object.keys(reducerDic)
        .forEach(action => {
          if (!output[action]) {
            output[action] = reducerDic[action];
            return;
          }

          const originalReducer = output[action];
          output[action] = (params) => {
            const stateA = params.state || {};
            const stateB = originalReducer({...params, state: stateA as any});
            return reducerDic[action]({
              ...params,
              state: {
                ...stateA,
                ...stateB,
              } as any,
            });
          };
        });
    });

  return output;
};
