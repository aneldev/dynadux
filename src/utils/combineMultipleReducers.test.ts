import "jest";
import {IDynaduxReducerDic} from "../index";
import {combineMultipleReducers} from "./combineMultipleReducers";

describe('Dynadux', () => {

  describe('combineMultipleReducers', () => {
    test('no duplicate actions', () => {
      const log: string[] = [];
      const products: IDynaduxReducerDic<any> = {
        LOGIN: () => log.push('products-LOGIN') as any,
        GET_PRODUCTS: () => log.push('products-GET_PRODUCTS') as any,
        LOGOUT: () => log.push('products-LOGOUT') as any,
      };
      const cart: IDynaduxReducerDic<any> = {
        LOGIN: () => log.push('cart-LOGIN') as any,
        ADD_PRODUCT: () => log.push('cart-ADD_PRODUCT') as any,
        REMOVE_PRODUCT: () => log.push('cart-REMOVE_PRODUCT') as any,
        LOGOUT: () => log.push('cart-LOGOUT') as any,
      };

      const allReducers = combineMultipleReducers(products, cart);

      allReducers['LOGIN']('reducerAPI' as any);
      allReducers['GET_PRODUCTS']('reducerAPI' as any);
      allReducers['ADD_PRODUCT']('reducerAPI' as any);
      allReducers['REMOVE_PRODUCT']('reducerAPI' as any);
      allReducers['LOGOUT']('reducerAPI' as any);

      expect(log).toMatchSnapshot();

    });
  });

});
