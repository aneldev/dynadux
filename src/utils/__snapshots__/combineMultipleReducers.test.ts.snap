// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynadux combineMultipleReducers no duplicate actions 1`] = `
Array [
  "products-LOGIN",
  "cart-LOGIN",
  "products-GET_PRODUCTS",
  "cart-ADD_PRODUCT",
  "cart-REMOVE_PRODUCT",
  "products-LOGOUT",
  "cart-LOGOUT",
]
`;
